voxel_grid/prototype/mvp_v2.hpp

```diff
@@ -0,0 +1,254 @@
+
+// structure of a single voxel
+struct Voxel {
+  Vec3f centroid = Vec3f::Zero();
+  size_t num_pt = 0;
+};
+
+// structure and operations of the grid
+template <typename PointT>
+class VoxelStructT {
+ public:
+  // Require user to define them
+  using PointCloud = PointCloudT<PointT>;
+  using Grid = std::unordered_map<size_t, Voxel>;
+
+  auto begin() { return grid_.begin(); }
+  auto end() { return grid_.end(); }
+  size_t size() { return grid_.size(); }
+
+ protected:
+  bool setUp(const TransformFilter<VoxelStructT> *transform_filter) {
+    const auto grid_filter =
+        static_cast<const GridFilterBase<VoxelStructT> *>(transform_filter);
+
+    Vec3f min_p, max_p;
+    const auto &input = grid_filter->getInputCloud();
+    grid_filter->getMinMax3D(input, min_p, max_p);
+
+    inverse_leaf_size_ = 1 / grid_filter->getLeafSize().array();
+    min_b_ = (min_p.array() * inverse_leaf_size_.array()).cast<int>();
+    max_b_ = (max_p.array() * inverse_leaf_size_.array()).cast<int>();
+
+    div_b_ = (max_b_ - min_b_).array() + 1;
+    divb_mul_ << div_b_[0], div_b_[0], div_b_[1];
+
+    // NOTE: we need to store the members variables of GridFilterBase which are
+    // needed by the member functions in this class as member variables
+    // e.g.
+    min_points_per_voxel_ = grid_filter->getMinimumPointsNumberPerVoxel();
+    // ...
+
+    // we have 10 points in main
+    leaf_layout_.resize(10, -1);
+
+    return true;
+  }
+
+  void saveResults(TransformFilter<VoxelStructT> *transform_filter) {
+    // COMMENT:
+    // 1. need to duplicate leaf_layout_ in both base and GridStruct
+    // 2. need to declare friend in GridFilterBase (for every grid filters which
+    // have additional results). Otherwise we have to declare a public
+    // setLeafLayout which doesn't make sense in the user's point of view
+
+    auto grid_filter =
+        static_cast<GridFilterBase<VoxelStructT> *>(transform_filter);
+    grid_filter->leaf_layout_ = std::move(leaf_layout_);
+  }
+
+  size_t hashPoint(const PointT &pt) {
+    const Vec3f tmp = pt.p.array() * inverse_leaf_size_.array();
+    return divb_mul_.dot(tmp.cast<int>() - min_b_);
+  }
+
+  void addPointToGrid(const PointT &pt) {
+    // can be batch/running mode
+    const size_t h = hashPoint(pt);
+    grid_[h].centroid += pt.p;
+    grid_[h].num_pt++;
+  }
+
+  boost::optional<PointT> filterGrid(Grid::iterator grid_it) {
+    const auto &voxel = grid_it->second;
+    if (voxel.num_pt >= min_points_per_voxel_) {
+      leaf_layout_[grid_it->first] = num_centroids++;
+      return PointT(voxel.centroid / voxel.num_pt);
+    } else {
+      leaf_layout_[grid_it->first] = -1;
+      return boost::none;
+    }
+  }
+
+  Eigen::Vector3i min_b_, max_b_, div_b_, divb_mul_;
+  Vec3f inverse_leaf_size_;
+  size_t min_points_per_voxel_;
+  Grid grid_;
+
+  size_t num_centroids = 0;
+  std::vector<int> leaf_layout_;
+
+ private:
+  template <typename, typename>
+  friend class TransformFilter;
```

Comment:
What hidden data do we want to access from transformFilter?

Reply:
We don't need it actually, I'll remove it and add functions for accessing data in `TransformFilter`